File: typescript/src/puzzles/puzzle-01-sober-savant.ts
import { BOTCModel } from "../model";
import { printSolution } from "../display";
import { KissatBackend, type SatBackend } from "../sat";
import {
  Chef,
  Drunk,
  Imp,
  Investigator,
  Knight,
  Noble,
  Savant,
  ScarletWoman,
  Seamstress,
  Steward,
  applyClaims,
  playerNames,
  script,
} from "../characters";
import { drunkBetweenTwoTownsfolk } from "../predicates";

export const PLAYERS = [
  new Investigator({ name: "Oscar", minionRole: ScarletWoman, among: ["Anna", "Sula"] }),
  new Noble({ name: "Matt", oneEvilAmong: ["Tim", "Oscar", "Sula"] }),
  new Seamstress({ name: "Anna", aligned: false, among: ["Oscar", "Sula"] }),
  new Savant({
    name: "You",
    statements: [
      (game) => [game.roleInPlay(Investigator), game.sitsNextToEvil("You")],
      (game) => [Chef.learnsCount(game, 1, "you_savant_chef", { registers: false }), drunkBetweenTwoTownsfolk(game)],
      (game) => [
        game.anyOf([game.isMinion("Tim"), game.isMinion("Sula")], "tim_or_sula_minion"),
        game.not(game.roleInPlay(Noble), "noble_not_in_play"),
      ],
    ],
  }),
  new Knight({ name: "Tim", noDemonAmong: ["Anna", "Sula"] }),
  new Steward({ name: "Sula", goodPlayer: "Matt" }),
];

export const PLAYER_NAMES = playerNames(PLAYERS);
export const CHARACTERS = script(Imp, ScarletWoman, Drunk, Investigator, Knight, Noble, Savant, Seamstress, Steward);

export function buildModel(backend: SatBackend): BOTCModel {
  const game = new BOTCModel(PLAYER_NAMES, { characters: CHARACTERS, seating: PLAYER_NAMES, backend });
  game.setCharacterCount(Imp, 1);
  game.setCharacterCount(ScarletWoman, 1);
  game.setCharacterCount(Drunk, 1);
  game.fixActual("You", Savant);
  applyClaims(game, PLAYERS);
  return game;
}

export async function solve() {
  return buildModel(await KissatBackend.create()).solveAll();
}

if (import.meta.main && process.argv[1]?.endsWith("puzzle-01-sober-savant.ts"))
  printSolution(await solve(), PLAYER_NAMES, { forcedRoles: [Imp, ScarletWoman, Drunk] });

File: puzzles/puzzle_01_sober_savant.py
'''
"Can the sober Savant solve the puzzle?"
by Not Quite Tangible

https://www.reddit.com/r/BloodOnTheClocktower/comments/1erb5e2/can_the_sober_savant_solve_the_puzzle/
'''

from __future__ import annotations

from botc_solver import BOTCModel, RoleClaim, print_solution
from botc_solver.characters import (
    DRUNK,
    IMP,
    INVESTIGATOR,
    KNIGHT,
    NOBLE,
    SAVANT,
    SCARLET_WOMAN,
    SEAMSTRESS,
    STEWARD,
    script,
)
from botc_solver.predicates import (
    chef_count_is,
    different_alignments,
    drunk_between_two_townsfolk,
    exactly_n_evil,
    sits_next_to_evil,
)


PLAYERS = ["Oscar", "Matt", "Anna", "You", "Tim", "Sula"]
CHARACTERS = script(
    IMP,
    SCARLET_WOMAN,
    DRUNK,
    INVESTIGATOR,
    KNIGHT,
    NOBLE,
    SAVANT,
    SEAMSTRESS,
    STEWARD,
)
CLAIMS = {
    "Oscar": "Investigator",
    "Matt": "Noble",
    "Anna": "Seamstress",
    "You": "Savant",
    "Tim": "Knight",
    "Sula": "Steward",
}


def build_model() -> BOTCModel:
    game = BOTCModel(PLAYERS, characters=CHARACTERS, seating=PLAYERS)

    game.set_character_count("Imp", 1)
    game.set_character_count("Scarlet Woman", 1)
    game.set_character_count("Drunk", 1)

    game.fix_actual("You", "Savant")
    for player in PLAYERS:
        game.fix_poisoned(player, False)
        game.set_apparent_role(player, CLAIMS[player])

    for player in PLAYERS:
        if player != "You":
            game.add_role_claim(RoleClaim(player, CLAIMS[player]))

    oscar_claim = game.any_of(
        [game.is_minion("Anna"), game.is_minion("Sula")],
        "oscar_claim_anna_or_sula_minion",
    )
    sula_claim = game.is_good("Matt")
    matt_claim = exactly_n_evil(game, ["Tim", "Oscar", "Sula"], 1)
    anna_claim = different_alignments(game, "Oscar", "Sula")
    tim_claim = game.not_(
        game.any_of(
            [game.is_demon("Anna"), game.is_demon("Sula")],
            "anna_or_sula_demon",
        ),
        "neither_anna_nor_sula_demon",
    )

    game.add_truthful_info_claim("Oscar", "Investigator", oscar_claim)
    game.add_truthful_info_claim("Sula", "Steward", sula_claim)
    game.add_truthful_info_claim("Matt", "Noble", matt_claim)
    game.add_truthful_info_claim("Anna", "Seamstress", anna_claim)
    game.add_truthful_info_claim("Tim", "Knight", tim_claim)

    game.add_exactly_one(
        [
            game.role_in_play("Investigator"),
            sits_next_to_evil(game, "You"),
        ]
    )
    game.add_exactly_one(
        [
            chef_count_is(game, 1),
            drunk_between_two_townsfolk(game),
        ]
    )
    game.add_exactly_one(
        [
            game.any_of(
                [game.is_minion("Tim"), game.is_minion("Sula")],
                "tim_or_sula_minion",
            ),
            game.not_(game.role_in_play("Noble"), "noble_not_in_play"),
        ]
    )

    return game


def solve():
    return build_model().solve_all()


def main() -> None:
    print_solution(solve(), PLAYERS, forced_roles=["Imp", "Scarlet Woman", "Drunk"])


if __name__ == "__main__":
    main()
